nodejs/private/ts_project.bzl

```diff
@@ -57,40 +23,10 @@ def _declare_outputs(ctx, paths):
         for path in paths
     ]
 
-def _calculate_root_dir(ctx):
-    some_generated_path = None
-    some_source_path = None
-    root_path = None
-
-    # Note we don't have access to the ts_project macro allow_js param here.
-    # For error-handling purposes, we can assume that any .js/.jsx
-    # input is meant to be in the rootDir alongside .ts/.tsx sources,
-    # whether the user meant for them to be sources or not.
-    # It's a non-breaking change to relax this constraint later, but would be
-    # a breaking change to restrict it further.
-    allow_js = True
-    for src in ctx.files.srcs:
-        if _lib.is_ts_src(src.path, allow_js):
-            if src.is_source:
-                some_source_path = src.path
-            else:
-                some_generated_path = src.path
-                root_path = ctx.bin_dir.path
-
-    if some_source_path and some_generated_path:
-        fail("ERROR: %s srcs cannot be a mix of generated files and source files " % ctx.label +
-             "since this would prevent giving a single rootDir to the TypeScript compiler\n" +
-             "    found generated file %s and source file %s" %
-             (some_generated_path, some_source_path))
-
-    return _lib.join(
-        root_path,
-        ctx.label.workspace_root,
-        ctx.label.package,
-        ctx.attr.root_dir,
-    )
-
-def _ts_project_impl(ctx):
+# This function has two extra arguments that are particular to how it's called locally.
+# - run_action is used with the build_bazel_rules_nodejs linker, but other rule authors might just use ctx.actions.run
```

Comment:
Hmm. `ctx.actions.run` wouldn't accept a `link_workspace_root` parameter

Reply:
yes that's true, it's compensated over here
https://github.com/aspect-build/rules_js/pull/27/files#diff-8bddb9b75b98520465cccb6b6633cb8e51d69f2b208047b6a4082b91e1440494R30

maybe there's a better way 🤔

Reply:
It would have to accept a `partial` so that the link_workspace_root argument to `run_node` could be applied before passing the helper into the library code. I think it's too complex to add that here...

Reply:
Ah, I found a better way - make the run_action helper default to None, so a TS rule implementation can just do

```
load("@rules_nodejs//nodejs/private:ts_project.bzl", _ts_project_lib = "ts_project")

ts_project = rule(
    implementation = _ts_project_lib.implementation,
    attrs = _ts_project_lib.attrs,
)
```

and when the `run_action` is not None, then we know we can pass the link_workspace_root as well.

Also moved link_workspace_root attr to be on the build_bazel_rules_nodejs implementation.